Move json_normalize to pd namespace

vishwakftw opened the pull request:
- Added a whatsnew entry
- Imported pandas.io.json.json_normalize in __init__.py

- [x] closes #27586 
- [x] whatsnew entry

cc: @WillAyd

vishwakftw committed:
Move json_normalize to pd namespace

- Added a whatsnew entry
- Imported pandas.io.json.json_normalize in __init__.py

WillAyd commented:
Makes sense for the move. @TomAugspurger do you know how we've messaged things like this to users in the past to not import from pandas.io.json anymore?

TomAugspurger commented:
I don't recall on messaging. Probably something "Accessing json_normalize from pandas.io has been deprecated and will be removed in a future version. Uses pandas.json_normalize instead."

But we'll need to make a `pandas.io._json_normalize` that has the implementation, and `pandas.io.json_normalize` will warn then call that. `pandas/__init__.py` will import `io._json_normalize as json_normalize`.

jreback commented on pandas/__init__.py:
```diff
@@ -178,6 +178,10 @@
     read_spss,
 )
 
+from pandas.io.json import (
```
you need to update the tests, we very explicity test what is in and not in pd

jreback commented on doc/source/whatsnew/v1.0.0.rst:
```diff
@@ -52,7 +52,7 @@ Other API changes
 Deprecations
 ~~~~~~~~~~~~
-
+- Moved :func:`pandas.io.json.json_normalize` to the top-level namespace as :func:`pandas.json_normalize`.
```
you didn't move, you exposed

jreback replied:
need to update any documentation that uses this in io.rst

vishwakftw commented:
Thanks for review @jreback, please allow me to get back on this by the end of the weekend if that’s ok. I’ll also incorporate @TomAugspurger’s comment on deprecation as well.

TomAugspurger commented:
Thanks! No rush, we’re still a ways out from 1.0. 

> On Jul 26, 2019, at 17:37, Vishwak Srinivasan <notifications@github.com> wrote:
> 
> Thanks for review @jreback, please allow me to get back on this by the end of the weekend if that’s ok. I’ll also incorporate @TomAugspurger’s comment on deprecation as well.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

vishwakftw committed:
Code review - I

- json_normalize --> _json_normalize
- json_normalize inside io.json._normalize is a deprecated method
- update whatsnew docs
- update io docs
- fix broken test_api

pep8speaks commented:
Hello @vishwakftw! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:





* In the file [`pandas/io/json/_normalize.py`](https://github.com/pandas-dev/pandas/blob/ea57d9845ff24b43a540cddb887e0a904d52608d/pandas/io/json/_normalize.py):

> [Line 348:13](https://github.com/pandas-dev/pandas/blob/ea57d9845ff24b43a540cddb887e0a904d52608d/pandas/io/json/_normalize.py#L348): [E126](https://duckduckgo.com/?q=pep8%20E126) continuation line over-indented for hanging indent
> [Line 352:13](https://github.com/pandas-dev/pandas/blob/ea57d9845ff24b43a540cddb887e0a904d52608d/pandas/io/json/_normalize.py#L352): [E123](https://duckduckgo.com/?q=pep8%20E123) closing bracket does not match indentation of opening bracket's line

vishwakftw committed:
Fix lint

vishwakftw committed:
Fix failing tests due to usage of deprecated function

vishwakftw committed:
Fix lint using black and isort, sorry

TomAugspurger commented:
Looks like the docs need to be updated: https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=15186

The line numbers may be a bit off, but hopefully close enough.

TomAugspurger commented on doc/source/whatsnew/v1.0.0.rst:
```diff
@@ -52,7 +52,7 @@ Other API changes
 Deprecations
 ~~~~~~~~~~~~
-
+- :func:`pandas.json_normalize` is now exposed in the top-level namespace. Usage of :func:`json_normalize` as :func:`pandas.io.json.json_normalize` is now deprecated and it is recommended to use :func:`json_normalize` as :func:`pandas.json_normalize` instead. (:issue:`27586`)
```
I would remove the :func:\`pandas.io.json.json_normalize\` as it's the deprecated one. Just use double-backticks instead.

vishwakftw replied:
Sure, thank you for the pointer.

vishwakftw committed:
Code Review - II

- Update whatsnew entry

vishwakftw committed:
Code Review - II

- Update whatsnew entry

TomAugspurger commented:
Can you update `doc/source/reference/io.rst` so that the reference docs for `json_normalize` are for the top-level?

It seems that `build_table_scheme` also is exposed. Would you be willing to do a similar change for it? Either here or as a followup.

vishwakftw commented:
> Can you update doc/source/reference/io.rst so that the reference docs for json_normalize are for the top-level?

Yes, will do that.

> It seems that build_table_scheme also is exposed. Would you be willing to do a similar change for it? Either here or as a followup.

Would you like me to expose `build_table_schema` in the pd namespace?

TomAugspurger commented:
Possibly. I didn't closely follow the issue discussing json_nomalize, but
we should be consistent. @jreback thoughts?

On Mon, Jul 29, 2019 at 9:54 AM Vishwak Srinivasan <notifications@github.com>
wrote:

> Can you update doc/source/reference/io.rst so that the reference docs for
> json_normalize are for the top-level?
>
> Yes, will do that.
>
> It seems that build_table_scheme also is exposed. Would you be willing to
> do a similar change for it? Either here or as a followup.
>
> Would you like me to expose build_table_schema in the pd namespace?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/27615?email_source=notifications&email_token=AAKAOIVMVRVD4WW6VZDGKPTQB4ADJA5CNFSM4IHFVMN2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD3A67OY#issuecomment-516026299>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAKAOIQ2DKANOQIP4S5GFA3QB4ADJANCNFSM4IHFVMNQ>
> .
>

vishwakftw committed:
Code Review - III

- Update entry in reference/io.rst

jorisvandenbossche commented:
I also just commented on the issue about `build_table_schema` (https://github.com/pandas-dev/pandas/issues/27586), but I am not sure that that warrants a top-level function.

WillAyd commented:
`build_table_schema` doesn't deserve a top level function as it's not generally useful for end users (it's one part of actually writing out in the table format) 

It's unfortunate that `json_normalize` leaked publicly the way it is now but it certainly has usage and enhancement requests coming in for it, hence this PR

TomAugspurger commented:
I wouldn't say "leaked publicly" though, right? That's the intentional,
documented location for it.

On Wed, Jul 31, 2019 at 4:06 PM William Ayd <notifications@github.com>
wrote:

> build_table_schema doesn't deserve a top level function as it's not
> generally useful for end users (it's one part of actually writing out in
> the table format)
>
> It's unfortunate that json_normalize leaked publicly the way it is now
> but it certainly has usage and enhancement requests coming in for it, hence
> this PR
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/27615?email_source=notifications&email_token=AAKAOIUX3GSP7YXWHME5HYLQCH5DTA5CNFSM4IHFVMN2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD3IR2NQ#issuecomment-517020982>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAKAOISWNHKHZYOLLMDUJ23QCH5DTANCNFSM4IHFVMNQ>
> .
>

WillAyd commented:
Can’t speak to the entire history of intention but looks like was added alongside read_json in #10301 but never moved to the top level whenever read_json was

In any case I don’t think whatever we do with this function needs to dictate what happens for build_table_schema

> On Jul 31, 2019, at 2:09 PM, Tom Augspurger <notifications@github.com> wrote:
> 
> I wouldn't say "leaked publicly" though, right? That's the intentional,
> documented location for it.
> 
> On Wed, Jul 31, 2019 at 4:06 PM William Ayd <notifications@github.com>
> wrote:
> 
> > build_table_schema doesn't deserve a top level function as it's not
> > generally useful for end users (it's one part of actually writing out in
> > the table format)
> >
> > It's unfortunate that json_normalize leaked publicly the way it is now
> > but it certainly has usage and enhancement requests coming in for it, hence
> > this PR
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/pandas-dev/pandas/pull/27615?email_source=notifications&email_token=AAKAOIUX3GSP7YXWHME5HYLQCH5DTA5CNFSM4IHFVMN2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD3IR2NQ#issuecomment-517020982>,
> > or mute the thread
> > <https://github.com/notifications/unsubscribe-auth/AAKAOISWNHKHZYOLLMDUJ23QCH5DTANCNFSM4IHFVMNQ>
> > .
> >
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/pandas-dev/pandas/pull/27615?email_source=notifications&email_token=AAEU4UIDSHL3FAQU6J7LRO3QCH5RZA5CNFSM4IHFVMN2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD3ISDQI#issuecomment-517022145>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AAEU4UL5H3YJORQL5ITYKOTQCH5RZANCNFSM4IHFVMNQ>.
>

vishwakftw committed:
Merge branch 'json_normalize-to-pd' of github.com:vishwakftw/pandas into json_normalize-to-pd

WillAyd commented:
@vishwakftw can you check doc failure and merge master?

vishwakftw committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into json_normalize-to-pd

vishwakftw committed:
Possible doc failure fix

vishwakftw commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -172,7 +172,7 @@ The repr now looks like this:
 .. ipython:: python
 
-    from pandas.io.json import json_normalize
+    from pandas import json_normalize
```
But this is causing the doc build failure, because there is a warning being thrown.

WillAyd replied:
Ah makes sense. Can you convert this into a `code-block` directive instead of `ipython` then? You should see examples of that in old whatsnew files that compare before / after behavior of changes

vishwakftw commented on doc/redirects.csv:
```diff
@@ -792,14 +792,14 @@ generated/pandas.io.formats.style.Styler.to_excel,../reference/api/pandas.io.for
 generated/pandas.io.formats.style.Styler.use,../reference/api/pandas.io.formats.style.Styler.use
 generated/pandas.io.formats.style.Styler.where,../reference/api/pandas.io.formats.style.Styler.where
 generated/pandas.io.json.build_table_schema,../reference/api/pandas.io.json.build_table_schema
-generated/pandas.io.json.json_normalize,../reference/api/pandas.io.json.json_normalize
```
Sure, would that be something like this:
```
generated/pandas.io.json.json_normalize,../reference/api/pandas.json_normalize
```
?

WillAyd replied:
Yea I think that works. If you have time to build and check would be great

vishwakftw commented on doc/source/user_guide/io.rst:
```diff
@@ -2153,7 +2153,7 @@ into a flat table.
 .. ipython:: python
 
-   from pandas.io.json import json_normalize
+   from pandas import json_normalize
```
Sure, I'll this out. Thank you.

vishwakftw committed:
Code Review - IV

vishwakftw committed:
Fix lint

vishwakftw commented:
@WillAyd The requested changes have been applied. Sorry about the delay in getting to them.

WillAyd commented on doc/source/user_guide/io.rst:
```diff
@@ -2153,27 +2153,26 @@ into a flat table.
-            'counties': [{'name': 'Dade', 'population': 12345},
-                         {'name': 'Broward', 'population': 40000},
-                         {'name': 'Palm Beach', 'population': 60000}]},
+            'county': [{'name': 'Dade', 'population': 12345},
```
Any reason for changing this? I think OK as is was?

vishwakftw replied:
I changed it because the example became longer, which caused the lint check to fail.

WillAyd commented on pandas/io/json/__init__.py:
```diff
@@ -1,12 +1,13 @@
     "loads",
     "read_json",
     "to_json",
+    "_json_normalize",
```
Does this need to be a part of `__all__`?

vishwakftw replied:
Yes, because we are aliasing it in `pandas/__init__.py` as `json_normalize`.

WillAyd commented on pandas/tests/io/json/test_normalize.py:
```diff
@@ -5,10 +5,9 @@
 from pandas.compat import PY36
 
-from pandas import DataFrame, Index
+from pandas import DataFrame, Index, json_normalize
```
Can you add a test here that importing from pandas.io.json now throws a `FutureWarning`?

vishwakftw replied:
Sure, I will do that.

vishwakftw committed:
Code Review - V

vishwakftw commented:
@WillAyd extremely sorry for not getting to the comments earlier, I've addressed your most recent comments.

vishwakftw committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into json_normalize-to-pd

WillAyd commented on pandas/tests/io/json/test_normalize.py:
```diff
@@ -692,3 +691,12 @@ def test_with_large_max_level(self):
+    def test_deprecated_import(self):
+        with tm.assert_produces_warning(FutureWarning) as w:
+            from pandas.io.json import json_normalize
+            recs = [
```
Is this on down necessary? I think should just throw warning on import not on invocation

vishwakftw replied:
I am not sure about how to deprecate while importing, do you have an idea?

WillAyd commented:
@vishwakftw is this still active? Can you merge master and repush?

vishwakftw committed:
Merge branch 'master' into json_normalize-to-pd

vishwakftw commented:
@WillAyd I was unsure about how to deprecate on import, and was wondering if you had an idea (https://github.com/pandas-dev/pandas/pull/27615/#discussion_r330086223).

WillAyd commented:
not sure; this might be OK as is if you can get CI to pass will take another look

vishwakftw committed:
Fix lint error

vishwakftw committed:
Merge branch 'master' into json_normalize-to-pd

vishwakftw commented:
I'm not sure if the CI failures are relevant.

WillAyd commented:
Looks like you need to run black on the changes to pass CI

I’ll take a look at the rest this weekend

vishwakftw committed:
Black reformat

vishwakftw committed:
Merge branch 'json_normalize-to-pd' of github.com:vishwakftw/pandas into json_normalize-to-pd

WillAyd reviewed (approved):
lgtm. Over to @jreback

vishwakftw committed:
Merge branch 'master' into json_normalize-to-pd

WillAyd commented:
@vishwakftw can you fix conflicts and ping on green

vishwakftw committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into json_normalize-to-pd

vishwakftw committed:
Merge branch 'json_normalize-to-pd' of github.com:vishwakftw/pandas into json_normalize-to-pd

vishwakftw commented:
@WillAyd Sure, I've rebased and will ping you once green.

vishwakftw commented:
@WillAyd There is one failure, I'm not sure if it is related:
```
=================================== FAILURES ===================================

______________________________ test_scikit_learn _______________________________

[gw1] linux -- Python 3.6.9 /home/travis/miniconda3/envs/pandas-dev/bin/python

df =    A

0  1

1  2

2  3

    @pytest.mark.filterwarnings("ignore:can't:ImportWarning")

    def test_scikit_learn(df):

    

>       sklearn = import_module("sklearn")  # noqa

pandas/tests/test_downstream.py:72: 

_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

pandas/tests/test_downstream.py:20: in import_module

    return importlib.import_module(name)

../../../miniconda3/envs/pandas-dev/lib/python3.6/importlib/__init__.py:126: in import_module

    return _bootstrap._gcd_import(name[level:], package, level)

<frozen importlib._bootstrap>:994: in _gcd_import

    ???

<frozen importlib._bootstrap>:971: in _find_and_load

    ???

<frozen importlib._bootstrap>:955: in _find_and_load_unlocked

    ???

<frozen importlib._bootstrap>:665: in _load_unlocked

    ???

<frozen importlib._bootstrap_external>:678: in exec_module

    ???

<frozen importlib._bootstrap>:219: in _call_with_frames_removed

    ???

../../../miniconda3/envs/pandas-dev/lib/python3.6/site-packages/sklearn/__init__.py:75: in <module>

    from .utils._show_versions import show_versions

_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

    """

    # License: BSD 3 clause

    

    import platform

    import sys

    import importlib

    

>   from ._openmp_helpers import _openmp_parallelism_enabled

E   ImportError: dlopen: cannot load any more object with static TLS

../../../miniconda3/envs/pandas-dev/lib/python3.6/site-packages/sklearn/utils/_show_versions.py:12: ImportError
```

WillAyd commented:
Should have been resolved with changes to CI yesterday. Looks like another merge conflict - if you clean up and repush should be green now I think

vishwakftw committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into json_normalize-to-pd

vishwakftw commented:
@WillAyd the CI is green now

WillAyd commented:
Great job with this @vishwakftw !

vishwakftw commented:
Thank you @WillAyd. Apologies for the delay in getting this merged.
